aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp

```diff
@@ -1978,78 +1985,53 @@ static void lu_factor_batched_magma(const Tensor& input, const Tensor& pivots, c
 }
 
 static void lu_factor(const Tensor& input, const Tensor& pivots, const Tensor& infos, bool compute_pivots) {
-  // MAGMA does not work with batch_size == 0.
-  // CuSolver does not work when the matrices have no elements
-  if (input.numel() == 0) {
-    // zero out the infos as it will have one element if the input is a matrix of size (0, 0)
-    infos.zero_();
-    return;
-  }
-
   auto batch_size = batchCount(input);
   (void) batch_size; // Silence unused warning in some builds
+  auto m = input.size(-2);
+  auto n = input.size(-1);
 
-
-#if AT_MAGMA_ENABLED()
   const auto lu_factor_magma = [batch_size](const Tensor& input, const Tensor& pivots, const Tensor& infos, const bool compute_pivots) {
     if (batch_size == 1) {
-        lu_factor_looped_magma(input, pivots, infos, compute_pivots);
+      lu_factor_looped_magma(input, pivots, infos, compute_pivots);
     } else {
-      // There is a bug in lu_factor_batched_magma in MAGMA < 2.5.2, see
-      // https://bitbucket.org/icl/magma/issues/13/getrf_batched-kernel-produces-nans-on
-      std::tuple<magma_int_t, magma_int_t, magma_int_t> version;
-      magma_version(&std::get<0>(version), &std::get<1>(version), &std::get<2>(version));
-      if (version >= std::make_tuple<magma_int_t, magma_int_t, magma_int_t>(2, 5, 2)) {
-        lu_factor_batched_magma(input, pivots, infos, compute_pivots);
-      } else {
-        lu_factor_looped_magma(input, pivots, infos, compute_pivots);
-      }
+      lu_factor_batched_magma(input, pivots, infos, compute_pivots);
     }
   };
-#endif
 
+  const auto preferred_backend = at::globalContext().linalgPreferredBackend();
+#ifdef USE_CUSOLVER
+  const auto lu_factor_cusolver = [batch_size, m, n](const Tensor& input, const Tensor& pivots, const Tensor& infos, bool compute_pivots) {
+    if (batch_size == 1 || m != n || m >= 512) {
+      lu_factor_looped_cusolver(input, pivots, infos, compute_pivots);
```

Comment:
Is there a test coverage for that?

Reply:
This is tested here https://github.com/pytorch/pytorch/blob/d0dc7cb7743825eeb4eff804b22ec37cc1d78cb7/test/test_linalg.py#L5254-L5255

Reply:
Case `m >= 512` is missing.

Reply:
As mentioned in https://github.com/pytorch/pytorch/pull/73878#issuecomment-1128939858, the first PR of this stack adds a test that does exercise all the paths.